src/CodeGen_PyTorch.cpp

```diff
@@ -0,0 +1,395 @@
+                }
+                if (modifier & halide_handle_cplusplus_type::Pointer) {
+                    oss << " *";
+                }
+            }
+        }
+    } else {
+        // This ends up using different type names than OpenCL does
+        // for the integer vector types. E.g. uint16x8_t rather than
+        // OpenCL's short8. Should be fine as CodeGen_C introduces
+        // typedefs for them and codegen always goes through this
+        // routine or its override in CodeGen_OpenCL to make the
+        // names. This may be the better bet as the typedefs are less
+        // likely to collide with built-in types (e.g. the OpenCL
+        // ones for a C compiler that decides to compile OpenCL).
+        // This code also supports arbitrary vector sizes where the
+        // OpenCL ones must be one of 2, 3, 4, 8, 16, which is too
+        // restrictive for already existing architectures.
+        switch (type.bits()) {
+        case 1:
+            // bool vectors are always emitted as uint8 in the C++ backend
+            if (type.is_vector()) {
+                oss << "uint8x" << type.lanes() << "_t";
+            } else {
+                oss << "bool";
+            }
+            break;
+        case 8: case 16: case 32: case 64:
+            if (type.is_uint()) {
+                oss << 'u';
+            }
+            oss << "int" << type.bits();
+            if (type.is_vector()) {
+                oss << "x" << type.lanes();
+            }
+            oss << "_t";
+            break;
+        default:
+            user_error << "Can't represent an integer with this many bits in C: " << type << "\n";
+        }
+    }
+    if (include_space && needs_space)
+        oss << " ";
+    return oss.str();
+}
+
+string type_to_pytorch_tensor(Type type, bool isCuda) {
+    return "at::Tensor";
+}
+
+} // ns anon
+
+
+CodeGen_PyTorch::CodeGen_PyTorch(ostream &s, Target t, OutputKind output_kind,
+    std::string cpp_header) :
+    IRPrinter(s), target(t), output_kind(output_kind), cpp_header(cpp_header)
+{
+  // TODO(mgharbi): header guard and header / implementation split
+  stream << "#include <torch/extension.h>\n";
+  // TODO(mgharbi): find a shallower integration with torch cuda, handle no GPU case
+  stream << "#include <HalideBuffer.h>\n\n";
+
+  // Conditionally add CUDA features to the pytorch helper
+  // if(target.has_feature(Target::CUDA)) {
+  //   stream << "#define HL_PT_CUDA\n";
+  // }
+  stream << "#include <HalidePytorchHelpers.h>\n";
+  if(target.has_feature(Target::CUDA)) {
+    stream << "#include <ATen/cuda/CUDAContext.h>\n";  
+    stream << "#include <HalidePytorchCudaHelpers.h>\n";
+    // stream << "#undef HL_PT_CUDA\n";
+  }
+
+  std::vector<std::string> header_path = split_string(cpp_header, "/");
+  std::string header = header_path.back();
+  
+
+  stream << "\n#include \"" << header << "\"\n\n";
+  stream << "using Halide::Runtime::Buffer;\n\n";
+}
+
+CodeGen_PyTorch::~CodeGen_PyTorch() {
```

Comment:
This would probably be better as `=default` in the .h file

Reply:
done. fixed in 25a87ad